x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushDownMvExpandPastProject.java

```diff
@@ -0,0 +1,112 @@
+import org.elasticsearch.xpack.esql.core.expression.NamedExpression;
+import org.elasticsearch.xpack.esql.plan.logical.Eval;
+import org.elasticsearch.xpack.esql.plan.logical.LogicalPlan;
+import org.elasticsearch.xpack.esql.plan.logical.MvExpand;
+import org.elasticsearch.xpack.esql.plan.logical.Project;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public final class PushDownMvExpandPastProject extends OptimizerRules.OptimizerRule<MvExpand> {
+    @Override
+    protected LogicalPlan rule(MvExpand mvExpand) {
+        if (mvExpand.child() instanceof Project pj) {
+            List<NamedExpression> projections = new ArrayList<>(pj.projections());
+
+            // Skip if any projection alias shadows an input field name, as injecting an Eval would cause
+            // duplicate output attributes. This can happen with aliases generated by ResolveUnionTypesInUnionAll.
+            // Example:
+            // MvExpand[salary{r}#168,salary{r}#175]
+            // \_Project[[$$salary$converted_to$keyword{r$}#178 AS salary#168]]
+            //   \_UnionAll[[salary{r}#174, $$salary$converted_to$keyword{r$}#178]]
+            Set<String> inputNames = pj.inputSet().stream().map(NamedExpression::name).collect(Collectors.toSet());
+            if (projections.stream().anyMatch(e -> e instanceof Alias alias && inputNames.contains(alias.toAttribute().name()))) {
+                return mvExpand;
+            }
+
+            // Find if the target is aliased in the project and create an alias with temporary names for it.
+            for (int i = 0; i < projections.size(); i++) {
+                NamedExpression projection = projections.get(i);
+                if (projection instanceof Alias alias) {
+                    if (alias.toAttribute().semanticEquals(mvExpand.target().toAttribute())) {
+                        // Check if the alias's original field (child) is referenced elsewhere in the projections.
+                        // If the original field is not referenced by any other projection or alias,
+                        // we don't need to inject an Eval to preserve it, and can safely resolve renames and push down.
+                        if (projections.stream()
+                            .anyMatch(
+                                ne -> ne.semanticEquals(alias.child())
+                                    || ne instanceof Alias as && as.child().semanticEquals(alias.child()) && as != alias
+                            ) == false) {
+                            // The alias's original field is not referenced elsewhere, no need to preserve it,
+                            mvExpand = PushDownUtils.resolveRenamesFromProject(mvExpand, pj);
+                            break;
+                        }
+
+                        // for query like: row a = 2 | eval b = a ｜ keep * | mv_expand b
+                        Alias aliasAlias = new Alias(
```

Comment:
Why do we create a temporary alias for the target field?

I'd expect we can just take the already existing alias from the Project and leave it behind in the Eval.

In your example
```
row a = 2 | eval b = a ｜ keep * | mv_expand b
```
the alias from the `eval b = a` can just be left in place (except that it needs to be stuffed into the new eval rather than being part of the projections).